src/main/java/teammates/storage/api/StudentsDb.java

```diff
@@ -410,4 +454,147 @@ StudentAttributes makeAttributes(CourseStudent entity) {
+        }
+
+        public List<StudentAttributes.UpdateOptions> getNonExistentStudents() {
+            return nonExistentStudents;
+        }
+
+        public List<StudentAttributes.UpdateOptions> getInvalidUpdates() {
+            return invalidUpdates;
+        }
+
+        /**
+         * Add {@link StudentAttributes.UpdateOptions} to the current transaction.
+         */
+        public void addUpdateOptions(List<StudentAttributes.UpdateOptions> updateOptionsList) {
+            this.updateOptionsList.addAll(updateOptionsList);
+        }
+
+        private void construct() {
+            for (StudentAttributes.UpdateOptions updateOptions : updateOptionsList) {
+                Assumption.assertNotNull(updateOptions);
+
+                CourseStudent student =
+                        studentsDb.getCourseStudentEntityForEmail(updateOptions.getCourseId(), updateOptions.getEmail());
+                if (student == null) {
+                    nonExistentStudents.add(updateOptions);
+                    return;
+                }
+                StudentUpdate studentUpdate = new StudentUpdate();
+                StudentAttributes newAttributes = studentsDb.makeAttributes(student);
+                studentUpdate.setBefore(newAttributes);
+                newAttributes.update(updateOptions);
+                newAttributes.sanitizeForSaving();
+                if (!newAttributes.isValid()) {
+                    invalidUpdates.add(updateOptions);
+                    return;
+                }
+
+                studentUpdate.setAfter(newAttributes);
+                updatedStudents.add(studentUpdate);
+
+                boolean isEmailChanged = !student.getEmail().equals(newAttributes.email);
+                if (isEmailChanged) {
+                    studentsToCreate.add(newAttributes);
+                    studentIDsToDelete.add(student.getUniqueId());
+                    return;
+                }
+
+                // update only if change
+                boolean hasSameAttributes =
+                        StudentsDb.hasSameValue(student.getName(), newAttributes.getName())
+                                && StudentsDb.hasSameValue(student.getLastName(), newAttributes.getLastName())
+                                && StudentsDb.hasSameValue(student.getComments(), newAttributes.getComments())
+                                && StudentsDb.hasSameValue(student.getGoogleId(), newAttributes.getGoogleId())
+                                && StudentsDb.hasSameValue(student.getTeamName(), newAttributes.getTeam())
+                                && StudentsDb.hasSameValue(student.getSectionName(), newAttributes.getSection());
+                if (hasSameAttributes) {
+                    log.info(String.format(
+                            OPTIMIZED_SAVING_POLICY_APPLIED,
+                            CourseStudent.class.getSimpleName(),
+                            updateOptions));
+                    return;
```

Comment:
If you `return` in a for loop, you are going to miss out on the other updates :) 

So you are not even updating anyone if the first student has same attributes.